// @antv/s2/esm/common/interface/basic.js
export var CellBorderPosition;
(function (CellBorderPosition) {
    CellBorderPosition["TOP"] = "TOP";
    CellBorderPosition["BOTTOM"] = "BOTTOM";
    CellBorderPosition["LEFT"] = "LEFT";
    CellBorderPosition["RIGHT"] = "RIGHT";
})(CellBorderPosition || (CellBorderPosition = {}));
/**
 * 类似 background-clip 属性: https://developer.mozilla.org/en-US/docs/Web/CSS/background-clip
 * 分为三种类型：
 * borderBox: 整个 cell 的范围
 * paddingBox: cell 去除 border 的范围
 * contentBox: cell 去除 (border + padding) 的范围
 * -------------------------------
 * |b|           padding         |
 * |o|  |---------------------|  |
 * |r|  |                     |  |
 * |d|  |                     |  |
 * |e|  |---------------------|  |
 * |r|           padding         |
 * -------------------------------
 * -------border-bottom-----------
 * -------------------------------
 */
export var CellClipBox;
(function (CellClipBox) {
    CellClipBox["BORDER_BOX"] = "borderBox";
    CellClipBox["PADDING_BOX"] = "paddingBox";
    CellClipBox["CONTENT_BOX"] = "contentBox";
})(CellClipBox || (CellClipBox = {}));
export var Aggregation;
(function (Aggregation) {
    Aggregation["SUM"] = "SUM";
    Aggregation["MIN"] = "MIN";
    Aggregation["MAX"] = "MAX";
    Aggregation["AVG"] = "AVG";
    Aggregation["COUNT"] = "COUNT";
})(Aggregation || (Aggregation = {}));
