// @antv/s2/lib/interaction/base-interaction/click/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RowColumnClick = exports.PreviewClick = exports.MergedCellClick = exports.HeaderCellLinkClick = exports.DataCellClick = exports.CornerCellClick = void 0;
var corner_cell_click_1 = require("./corner-cell-click");
Object.defineProperty(exports, "CornerCellClick", { enumerable: true, get: function () { return corner_cell_click_1.CornerCellClick; } });
var data_cell_click_1 = require("./data-cell-click");
Object.defineProperty(exports, "DataCellClick", { enumerable: true, get: function () { return data_cell_click_1.DataCellClick; } });
var header_cell_link_click_1 = require("./header-cell-link-click");
Object.defineProperty(exports, "HeaderCellLinkClick", { enumerable: true, get: function () { return header_cell_link_click_1.HeaderCellLinkClick; } });
var merged_cell_click_1 = require("./merged-cell-click");
Object.defineProperty(exports, "MergedCellClick", { enumerable: true, get: function () { return merged_cell_click_1.MergedCellClick; } });
var preview_click_1 = require("./preview-click");
Object.defineProperty(exports, "PreviewClick", { enumerable: true, get: function () { return preview_click_1.PreviewClick; } });
var row_column_click_1 = require("./row-column-click");
Object.defineProperty(exports, "RowColumnClick", { enumerable: true, get: function () { return row_column_click_1.RowColumnClick; } });
