// @antv/s2/lib/extends/pivot-chart/header/axis-col.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AxisColHeader = void 0;
const g_1 = require("@antv/g");
const s2_1 = require("@antv/s2");
const axis_col_cell_1 = require("../cell/axis-col-cell");
const constant_1 = require("../constant");
const frozen_1 = require("../utils/frozen");
class AxisColHeader extends s2_1.ColHeader {
    initGroups() {
        this.scrollGroup = this.appendChild(new g_1.Group({
            name: constant_1.KEY_GROUP_COL_AXIS_SCROLL,
            style: { zIndex: s2_1.FRONT_GROUND_GROUP_SCROLL_Z_INDEX },
        }));
        this.frozenGroup = this.appendChild(new g_1.Group({
            name: constant_1.KEY_GROUP_COL_AXIS_FROZEN,
            style: { zIndex: s2_1.FRONT_GROUND_GROUP_FROZEN_Z_INDEX },
        }));
        this.frozenTrailingGroup = this.appendChild(new g_1.Group({
            name: constant_1.KEY_GROUP_COL_AXIS_FROZEN,
            style: { zIndex: s2_1.FRONT_GROUND_GROUP_FROZEN_Z_INDEX },
        }));
        const { spreadsheet, nodes } = this.getHeaderConfig();
        this.extraFrozenNodes = (0, frozen_1.getExtraFrozenColAxisNodes)(spreadsheet.facet, nodes);
    }
    getCellInstance(node) {
        const headerConfig = this.getHeaderConfig();
        const { spreadsheet } = headerConfig;
        const { axisColCell: colAxisCell } = spreadsheet.options;
        return ((colAxisCell === null || colAxisCell === void 0 ? void 0 : colAxisCell(node, spreadsheet, headerConfig)) ||
            new axis_col_cell_1.AxisColCell(node, spreadsheet, headerConfig));
    }
}
exports.AxisColHeader = AxisColHeader;
