// @antv/s2/lib/extends/pivot-chart/header/corner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CornerHeader = void 0;
const s2_1 = require("@antv/s2");
class CornerHeader extends s2_1.CornerHeader {
    static getCornerNodes(options) {
        var _a, _b, _c, _d, _e;
        const cornerNodes = super.getCornerNodes(options);
        // 创建角头区域竖轴
        const { seriesNumberWidth, layoutResult, spreadsheet } = options;
        const { rowsHierarchy, axisRowsHierarchy, colsHierarchy } = layoutResult;
        const rowAxisNode = axisRowsHierarchy === null || axisRowsHierarchy === void 0 ? void 0 : axisRowsHierarchy.sampleNodeForLastLevel;
        if (rowAxisNode) {
            const leafNode = colsHierarchy === null || colsHierarchy === void 0 ? void 0 : colsHierarchy.sampleNodeForLastLevel;
            const cornerNode = new s2_1.Node({
                id: rowAxisNode.id,
                field: rowAxisNode.field,
                value: spreadsheet.dataSet.getFieldName(rowAxisNode.field),
                x: seriesNumberWidth + rowsHierarchy.width + rowAxisNode.x,
                y: (_a = leafNode === null || leafNode === void 0 ? void 0 : leafNode.y) !== null && _a !== void 0 ? _a : 0,
                width: rowAxisNode.width,
                height: (_e = (_b = leafNode === null || leafNode === void 0 ? void 0 : leafNode.height) !== null && _b !== void 0 ? _b : spreadsheet.facet.getCellCustomSize(null, (_d = (_c = spreadsheet.options.style) === null || _c === void 0 ? void 0 : _c.colCell) === null || _d === void 0 ? void 0 : _d.height)) !== null && _e !== void 0 ? _e : 0,
                isPivotMode: true,
                cornerType: s2_1.CornerNodeType.Row,
                spreadsheet,
            });
            cornerNodes.push(cornerNode);
        }
        return cornerNodes;
    }
}
exports.CornerHeader = CornerHeader;
