// @antv/s2-react/lib/components/sheets/grid-analysis-sheet/theme.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GridAnalysisTheme = exports.FONT_SIZE_MINOR = exports.FONT_SIZE = void 0;
const s2_1 = require("@antv/s2");
exports.FONT_SIZE = 12;
exports.FONT_SIZE_MINOR = 11;
/* 网格分析表主题 */
exports.GridAnalysisTheme = {
    // 表头
    dataCell: {
        text: {
            fontSize: exports.FONT_SIZE,
            fontFamily: s2_1.FONT_FAMILY,
            fill: 'rgba(0, 0, 0, 0.85)',
            fontWeight: 'normal',
            textBaseline: 'middle',
            textAlign: 'left',
        },
        // 次级文本，如副指标
        minorText: {
            fontSize: exports.FONT_SIZE_MINOR,
            fontFamily: s2_1.FONT_FAMILY,
            fill: 'rgba(0, 0, 0, 0.45)',
            fontWeight: 'normal',
            textBaseline: 'middle',
            textAlign: 'left',
        },
        // 衍生指标
        derivedMeasureText: {
            mainUp: '#F46649',
            mainDown: '#2AA491',
            minorUp: '#f9ae9e',
            minorDown: '#a9dad2',
        },
        bolderText: {
            fontSize: exports.FONT_SIZE,
            fontFamily: s2_1.FONT_FAMILY,
            fill: 'rgba(0, 0, 0, 1)',
            fontWeight: (0, s2_1.isWindows)() ? 'bold' : 520,
            textBaseline: 'middle',
            textAlign: 'left',
        },
    },
};
