// @antv/s2/lib/extends/pivot-chart/constant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.X_FIELD_FORMATTER = exports.PLACEHOLDER_FIELD = exports.KEY_GROUP_COL_AXIS_RESIZE_AREA = exports.KEY_GROUP_COL_AXIS_FROZEN_TRAILING = exports.KEY_GROUP_COL_AXIS_FROZEN = exports.KEY_GROUP_COL_AXIS_SCROLL = exports.KEY_GROUP_ROW_AXIS_RESIZE_AREA = exports.KEY_GROUP_ROW_AXIS_HEADER_FROZEN_TRAILING = exports.KEY_GROUP_ROW_AXIS_FROZEN = exports.KEY_GROUP_ROW_AXIS_SCROLL = exports.DEFAULT_DIMENSION_SIZE = exports.DEFAULT_COL_AXIS_SIZE = exports.DEFAULT_ROW_AXIS_SIZE = exports.DEFAULT_MEASURE_SIZE = exports.FIXED_DATA_CONFIG = exports.DEFAULT_OPTIONS = exports.FIXED_OPTIONS = exports.DEFAULT_CHART_SPEC = exports.DEFAULT_G2_SPEC = void 0;
const axis_click_1 = require("./interaction/axis-click");
const axis_hover_1 = require("./interaction/axis-hover");
exports.DEFAULT_G2_SPEC = {
    autoFit: true,
    animate: false,
    // https://g2.antv.antgroup.com/manual/core/size
    margin: 1,
};
exports.DEFAULT_CHART_SPEC = Object.assign(Object.assign({}, exports.DEFAULT_G2_SPEC), { type: 'interval', axis: false, legend: false });
exports.FIXED_OPTIONS = {
    hierarchyType: 'grid',
    interaction: {
        selectedCellsSpotlight: false,
        copy: {
            enable: false,
        },
    },
    style: {
        colCell: {
            hideValue: false,
        },
    },
};
exports.DEFAULT_OPTIONS = {
    chart: {
        coordinate: 'cartesian',
    },
    interaction: {
        customInteractions: [
            {
                key: 'axisHover',
                interaction: axis_hover_1.AxisHover,
            },
            {
                key: 'axisClick',
                interaction: axis_click_1.AxisRowColumnClick,
            },
        ],
    },
};
exports.FIXED_DATA_CONFIG = {
    fields: {
        customValueOrder: null,
    },
};
exports.DEFAULT_MEASURE_SIZE = 200;
exports.DEFAULT_ROW_AXIS_SIZE = 100;
exports.DEFAULT_COL_AXIS_SIZE = 50;
exports.DEFAULT_DIMENSION_SIZE = 50;
/**
 * row axis
 */
exports.KEY_GROUP_ROW_AXIS_SCROLL = 'rowAxisScrollGroup';
exports.KEY_GROUP_ROW_AXIS_FROZEN = 'rowAxisHeaderFrozenGroup';
exports.KEY_GROUP_ROW_AXIS_HEADER_FROZEN_TRAILING = 'rowAxisHeaderFrozenTrailingGroup';
exports.KEY_GROUP_ROW_AXIS_RESIZE_AREA = 'rowAxisHeaderResizeArea';
/**
 * column axis
 */
exports.KEY_GROUP_COL_AXIS_SCROLL = 'colAxisScrollGroup';
exports.KEY_GROUP_COL_AXIS_FROZEN = 'colAxisFrozenGroup';
exports.KEY_GROUP_COL_AXIS_FROZEN_TRAILING = 'colAxisFrozenTrailingGroup';
exports.KEY_GROUP_COL_AXIS_RESIZE_AREA = 'colAxisHeaderResizeArea';
exports.PLACEHOLDER_FIELD = '$$placeholder$$';
exports.X_FIELD_FORMATTER = '$$should_formatter$$';
