// @antv/s2/lib/common/constant/events/origin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OriginEventType = void 0;
/**
 * @see https://g.antv.antgroup.com/api/event/intro
 */
var OriginEventType;
(function (OriginEventType) {
    OriginEventType["MOUSE_DOWN"] = "mousedown";
    OriginEventType["MOUSE_MOVE"] = "mousemove";
    OriginEventType["MOUSE_UP"] = "mouseup";
    OriginEventType["MOUSE_OUT"] = "mouseout";
    OriginEventType["MOUSE_LEAVE"] = "mouseleave";
    OriginEventType["TOUCH_START"] = "touchstart";
    OriginEventType["TOUCH_END"] = "touchend";
    OriginEventType["KEY_DOWN"] = "keydown";
    OriginEventType["KEY_UP"] = "keyup";
    OriginEventType["CLICK"] = "click";
    OriginEventType["HOVER"] = "hover";
    OriginEventType["CONTEXT_MENU"] = "contextmenu";
    OriginEventType["POINTER_DOWN"] = "pointerdown";
    OriginEventType["POINTER_MOVE"] = "pointermove";
    OriginEventType["POINTER_OUT"] = "pointerout";
    OriginEventType["POINTER_LEAVE"] = "pointerleave";
    OriginEventType["POINTER_UP"] = "pointerup";
    OriginEventType["POINTER_CANCEL"] = "pointercancel";
    OriginEventType["POINTER_ENTER"] = "pointerenter";
    OriginEventType["POINTER_OVER"] = "pointerover";
    OriginEventType["RIGHT_DOWN"] = "rightdown";
    OriginEventType["RIGHT_UP"] = "rightup";
    OriginEventType["SCROLL"] = "scroll";
})(OriginEventType || (exports.OriginEventType = OriginEventType = {}));
