// @antv/s2/lib/extends/pivot-chart/cell/axis-corner-cell.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AxisCornerCell = void 0;
const s2_1 = require("@antv/s2");
const cell_type_1 = require("./cell-type");
class AxisCornerCell extends s2_1.CornerCell {
    get cellType() {
        return cell_type_1.AxisCellType.AXIS_CORNER_CELL;
    }
    getBorderPositions() {
        return [
            s2_1.CellBorderPosition.TOP,
            s2_1.CellBorderPosition.BOTTOM,
            s2_1.CellBorderPosition.LEFT,
        ];
    }
    isBolderText() {
        return false;
    }
    getMaxTextWidth() {
        const { width } = this.getBBoxByType(s2_1.CellClipBox.CONTENT_BOX);
        return width;
    }
    getTreeIconWidth() {
        return 0;
    }
    getInteractedCells() {
        var _a;
        return (_a = this.spreadsheet.interaction) === null || _a === void 0 ? void 0 : _a.getCells([
            cell_type_1.AxisCellType.AXIS_CORNER_CELL,
        ]);
    }
    initCell() {
        this.drawBackgroundShape();
        this.drawTextShape();
        this.drawBorders();
        this.update();
    }
}
exports.AxisCornerCell = AxisCornerCell;
