// @antv/s2/lib/utils/export/copy/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ricOptions = exports.getHeaderNodeFromMeta = exports.getNodeFormatData = exports.unifyConfig = exports.getFormatOptions = exports.completeMatrix = exports.getMaxRowLen = exports.assembleMatrix = exports.Transformers = exports.matrixHtmlTransformer = exports.matrixPlainTextTransformer = void 0;
const lodash_1 = require("lodash");
const common_1 = require("../../../common");
const export_1 = require("../../../common/interface/export");
// 把 string[][] 矩阵转换成 CopyablePlain
const matrixPlainTextTransformer = (dataMatrix, separator = common_1.TAB_SEPARATOR) => {
    return {
        type: export_1.CopyMIMEType.PLAIN,
        content: (0, lodash_1.map)(dataMatrix, (line) => line.join(separator)).join(common_1.LINE_SEPARATOR),
    };
};
exports.matrixPlainTextTransformer = matrixPlainTextTransformer;
// 把 string[][] 矩阵转换成 CopyableItem
const matrixHtmlTransformer = (dataMatrix) => {
    function createTableData(data, tagName) {
        return data
            .map((cell) => `<${tagName}>${(0, lodash_1.escape)(cell)}</${tagName}>`)
            .join('');
    }
    function createBody(data, tagName) {
        return data
            .map((row) => `<${tagName}>${createTableData(row, 'td')}</${tagName}>`)
            .join('');
    }
    const body = createBody(dataMatrix, 'tr');
    return {
        type: export_1.CopyMIMEType.HTML,
        content: `<meta charset="utf-8"><table><tbody>${body}</tbody></table>`,
    };
};
exports.matrixHtmlTransformer = matrixHtmlTransformer;
exports.Transformers = {
    [export_1.CopyMIMEType.PLAIN]: exports.matrixPlainTextTransformer,
    [export_1.CopyMIMEType.HTML]: exports.matrixHtmlTransformer,
};
// 生成矩阵：https://gw.alipayobjects.com/zos/antfincdn/bxBVt0nXx/a182c1d4-81bf-469f-b868-8b2e29acfc5f.png
const assembleMatrix = ({ rowMatrix, colMatrix, dataMatrix, cornerMatrix, }) => {
    var _a, _b, _c, _d, _e, _f;
    const rowWidth = (_b = (_a = rowMatrix === null || rowMatrix === void 0 ? void 0 : rowMatrix[0]) === null || _a === void 0 ? void 0 : _a.length) !== null && _b !== void 0 ? _b : 0;
    const colHeight = (_c = colMatrix === null || colMatrix === void 0 ? void 0 : colMatrix.length) !== null && _c !== void 0 ? _c : 0;
    const dataWidth = (_e = (_d = dataMatrix[0]) === null || _d === void 0 ? void 0 : _d.length) !== null && _e !== void 0 ? _e : 0;
    const dataHeight = (_f = dataMatrix.length) !== null && _f !== void 0 ? _f : 0;
    const matrixWidth = rowWidth + dataWidth;
    // 存在行头节点时，则必然存在行头标题的角头
    const cornerHeight = colHeight || (rowWidth > 0 ? 1 : 0);
    const matrixHeight = cornerHeight + dataHeight;
    let matrix = Array.from(Array(matrixHeight), () => new Array(matrixWidth));
    matrix = (0, lodash_1.map)(matrix, (heightArr, y) => (0, lodash_1.map)(heightArr, (_, x) => {
        var _a, _b;
        if (x >= 0 && x < rowWidth && y >= 0 && y < cornerHeight) {
            return (_b = (_a = cornerMatrix === null || cornerMatrix === void 0 ? void 0 : cornerMatrix[y]) === null || _a === void 0 ? void 0 : _a[x]) !== null && _b !== void 0 ? _b : '';
        }
        if (x >= rowWidth && x <= matrixWidth && y >= 0 && y < colHeight) {
            return colMatrix[y][x - rowWidth];
        }
        if (x >= 0 && x < rowWidth && y >= cornerHeight && y < matrixHeight) {
            return rowMatrix === null || rowMatrix === void 0 ? void 0 : rowMatrix[y - cornerHeight][x];
        }
        if (x >= rowWidth &&
            x <= matrixWidth &&
            y >= cornerHeight &&
            y < matrixHeight) {
            return dataMatrix[y - cornerHeight][x - rowWidth];
        }
        return undefined;
    }));
    return matrix;
};
exports.assembleMatrix = assembleMatrix;
function getMaxRowLen(matrix) {
    var _a;
    return (_a = (0, lodash_1.max)((0, lodash_1.map)(matrix, (row) => row.length))) !== null && _a !== void 0 ? _a : 0;
}
exports.getMaxRowLen = getMaxRowLen;
/**
 * 补全 matrix 中的元素个数, 使得每一行的元素个数一致，以最大的行元素个数为准
 * @param {SimpleData[][]} matrix
 * @return {SimpleData[][]}
 */
function completeMatrix(matrix) {
    const maxRowLen = getMaxRowLen(matrix);
    return (0, lodash_1.map)(matrix, (row) => {
        const diff = maxRowLen - row.length;
        return diff ? [...row, ...new Array(diff).fill('')] : row;
    });
}
exports.completeMatrix = completeMatrix;
const getFormatOptions = (formatOptions) => {
    var _a, _b;
    if (typeof formatOptions === 'object') {
        return {
            formatHeader: (_a = formatOptions.formatHeader) !== null && _a !== void 0 ? _a : false,
            formatData: (_b = formatOptions.formatData) !== null && _b !== void 0 ? _b : false,
        };
    }
    return {
        formatHeader: formatOptions !== null && formatOptions !== void 0 ? formatOptions : false,
        formatData: formatOptions !== null && formatOptions !== void 0 ? formatOptions : false,
    };
};
exports.getFormatOptions = getFormatOptions;
function getTransformer(customTransformer) {
    if (!customTransformer) {
        return exports.Transformers;
    }
    const customTransformersTemp = customTransformer(exports.Transformers);
    return {
        [export_1.CopyMIMEType.PLAIN]: customTransformersTemp[export_1.CopyMIMEType.PLAIN] ||
            exports.Transformers[export_1.CopyMIMEType.PLAIN],
        [export_1.CopyMIMEType.HTML]: customTransformersTemp[export_1.CopyMIMEType.HTML] ||
            exports.Transformers[export_1.CopyMIMEType.HTML],
    };
}
// 因为 copy 和 export 在配置上有一定差异，此方法用于抹平差异
function unifyConfig({ spreadsheet: { options: { interaction }, }, config: { formatOptions = false, separator = common_1.TAB_SEPARATOR, selectedCells = [], customTransformer, async = false, }, isExport, }) {
    var _a, _b;
    const { copy } = interaction;
    const { formatData, formatHeader } = isExport
        ? (0, exports.getFormatOptions)(formatOptions)
        : {
            formatData: (_a = copy === null || copy === void 0 ? void 0 : copy.withFormat) !== null && _a !== void 0 ? _a : false,
            formatHeader: (_b = copy === null || copy === void 0 ? void 0 : copy.withFormat) !== null && _b !== void 0 ? _b : false,
        };
    const transformers = getTransformer(isExport ? customTransformer : copy === null || copy === void 0 ? void 0 : copy.customTransformer);
    return {
        separator,
        selectedCells,
        transformers,
        formatData,
        formatHeader,
        async,
    };
}
exports.unifyConfig = unifyConfig;
/**
 * 获取到当前节点所有数据
 * @param leafNode
 */
const getNodeFormatData = (leafNode) => {
    const line = [];
    const getNodeFormatterLabel = (node) => {
        var _a, _b, _c;
        // node.id === KEY_ROOT_NODE 时，为 S2 内的虚拟根节点，导出的内容不需要考虑此节点
        if (node.id === common_1.ROOT_NODE_ID) {
            return;
        }
        const formatter = (_c = (_b = (_a = node.spreadsheet) === null || _a === void 0 ? void 0 : _a.dataSet) === null || _b === void 0 ? void 0 : _b.getFieldFormatter) === null || _c === void 0 ? void 0 : _c.call(_b, node.field);
        const value = node.isTotalRoot || !formatter
            ? node.value
            : formatter(node.value);
        line.unshift(value);
        if (node === null || node === void 0 ? void 0 : node.parent) {
            return getNodeFormatterLabel(node.parent);
        }
    };
    getNodeFormatterLabel(leafNode);
    return line;
};
exports.getNodeFormatData = getNodeFormatData;
const getHeaderNodeFromMeta = (meta, spreadsheet) => {
    const { rowIndex, colIndex } = meta;
    const { facet } = spreadsheet;
    return [facet.getRowNodeByIndex(rowIndex), facet.getColNodeByIndex(colIndex)];
};
exports.getHeaderNodeFromMeta = getHeaderNodeFromMeta;
exports.ricOptions = { timeout: 3000 };
