// @antv/s2/lib/utils/export/copy/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.strategyCopy = exports.PivotDataCellCopy = exports.getSelectedData = exports.asyncGetAllPlainData = exports.asyncGetAllHtmlData = exports.asyncGetAllData = void 0;
var core_1 = require("./core");
Object.defineProperty(exports, "asyncGetAllData", { enumerable: true, get: function () { return core_1.asyncGetAllData; } });
Object.defineProperty(exports, "asyncGetAllHtmlData", { enumerable: true, get: function () { return core_1.asyncGetAllHtmlData; } });
Object.defineProperty(exports, "asyncGetAllPlainData", { enumerable: true, get: function () { return core_1.asyncGetAllPlainData; } });
Object.defineProperty(exports, "getSelectedData", { enumerable: true, get: function () { return core_1.getSelectedData; } });
var pivot_data_cell_copy_1 = require("./pivot-data-cell-copy");
Object.defineProperty(exports, "PivotDataCellCopy", { enumerable: true, get: function () { return pivot_data_cell_copy_1.PivotDataCellCopy; } });
var strategy_copy_1 = require("./strategy-copy");
Object.defineProperty(exports, "strategyCopy", { enumerable: true, get: function () { return strategy_copy_1.strategyCopy; } });
