// @antv/s2/esm/common/constant/frozen.js
export var FrozenGroupArea;
(function (FrozenGroupArea) {
    FrozenGroupArea["Row"] = "frozenRow";
    FrozenGroupArea["Col"] = "frozenCol";
    FrozenGroupArea["TrailingRow"] = "frozenTrailingRow";
    FrozenGroupArea["TrailingCol"] = "frozenTrailingCol";
})(FrozenGroupArea || (FrozenGroupArea = {}));
export var FrozenGroupType;
(function (FrozenGroupType) {
    FrozenGroupType["Row"] = "frozenRowGroup";
    FrozenGroupType["Col"] = "frozenColGroup";
    FrozenGroupType["TrailingCol"] = "frozenTrailingColGroup";
    FrozenGroupType["TrailingRow"] = "frozenTrailingRowGroup";
    FrozenGroupType["TopLeft"] = "frozenTopLeftGroup";
    FrozenGroupType["TopRight"] = "frozenTopRightGroup";
    FrozenGroupType["BottomLeft"] = "frozenBottomLeftGroup";
    FrozenGroupType["BottomRight"] = "frozenBottomRightGroup";
    FrozenGroupType["Scroll"] = "scrollGroup";
})(FrozenGroupType || (FrozenGroupType = {}));
export const FrozenGroupAreaTypeMap = {
    [FrozenGroupArea.Row]: FrozenGroupType.Row,
    [FrozenGroupArea.Col]: FrozenGroupType.Col,
    [FrozenGroupArea.TrailingRow]: FrozenGroupType.TrailingRow,
    [FrozenGroupArea.TrailingCol]: FrozenGroupType.TrailingCol,
};
