// @antv/s2/esm/common/constant/interaction.js
export var InteractionName;
(function (InteractionName) {
    InteractionName["CORNER_CELL_CLICK"] = "cornerCellClick";
    InteractionName["DATA_CELL_CLICK"] = "dataCellClick";
    InteractionName["ROW_CELL_CLICK"] = "rowCellClick";
    InteractionName["COL_CELL_CLICK"] = "colCellClick";
    InteractionName["MERGED_CELLS_CLICK"] = "mergedCellsClick";
    InteractionName["ROW_COLUMN_CLICK"] = "rowColumnClick";
    InteractionName["HEADER_CELL_LINK_CLICK"] = "headerCellLinkClick";
    InteractionName["PREVIEW_CLICK"] = "previewClick";
    InteractionName["HOVER"] = "hover";
    InteractionName["DATA_CELL_BRUSH_SELECTION"] = "dataCellBrushSelection";
    InteractionName["ROW_CELL_BRUSH_SELECTION"] = "rowCellBrushSelection";
    InteractionName["COL_CELL_BRUSH_SELECTION"] = "colCellBrushSelection";
    InteractionName["COL_ROW_RESIZE"] = "rowColResize";
    InteractionName["DATA_CELL_MULTI_SELECTION"] = "dataCellMultiSelection";
    InteractionName["ROW_CELL_MULTI_SELECTION"] = "rowCellMultiSelection";
    InteractionName["COL_CELL_MULTI_SELECTION"] = "colCellMultiSelection";
    InteractionName["RANGE_SELECTION"] = "rangeSelection";
    InteractionName["SELECTED_CELL_MOVE"] = "selectedCellMove";
    InteractionName["GLOBAL_RESET"] = "globalReset";
})(InteractionName || (InteractionName = {}));
export var InteractionStateName;
(function (InteractionStateName) {
    InteractionStateName["ALL_SELECTED"] = "allSelected";
    InteractionStateName["SELECTED"] = "selected";
    InteractionStateName["ROW_CELL_BRUSH_SELECTED"] = "rowCellBrushSelected";
    InteractionStateName["COL_CELL_BRUSH_SELECTED"] = "colCellBrushSelected";
    InteractionStateName["DATA_CELL_BRUSH_SELECTED"] = "dataCellBrushSelected";
    InteractionStateName["UNSELECTED"] = "unselected";
    InteractionStateName["HOVER"] = "hover";
    InteractionStateName["HOVER_FOCUS"] = "hoverFocus";
    InteractionStateName["HIGHLIGHT"] = "highlight";
    InteractionStateName["SEARCH_RESULT"] = "searchResult";
    InteractionStateName["PREPARE_SELECT"] = "prepareSelect";
})(InteractionStateName || (InteractionStateName = {}));
export var CellType;
(function (CellType) {
    CellType["DATA_CELL"] = "dataCell";
    CellType["ROW_CELL"] = "rowCell";
    CellType["COL_CELL"] = "colCell";
    CellType["SERIES_NUMBER_CELL"] = "seriesNumberCell";
    CellType["CORNER_CELL"] = "cornerCell";
    CellType["MERGED_CELL"] = "mergedCell";
})(CellType || (CellType = {}));
export const HOVER_FOCUS_DURATION = 800;
// 主题配置和canvas属性的映射
export const SHAPE_STYLE_MAP = {
    textOpacity: 'fillOpacity',
    backgroundOpacity: 'fillOpacity',
    backgroundColor: 'fill',
    borderOpacity: 'strokeOpacity',
    borderColor: 'stroke',
    borderWidth: 'lineWidth',
    opacity: 'opacity',
};
// 设置属性的时候实际对应改变的shape映射
export const SHAPE_ATTRS_MAP = {
    textShape: ['textOpacity'],
    textShapes: ['textOpacity'],
    linkFieldShape: ['opacity'],
    interactiveBgShape: ['backgroundColor', 'backgroundOpacity'],
    interactiveBorderShape: ['borderColor', 'borderOpacity', 'borderWidth'],
};
export const INTERACTION_STATE_INFO_KEY = 'interactionStateInfo';
export var InteractionBrushSelectionStage;
(function (InteractionBrushSelectionStage) {
    InteractionBrushSelectionStage["CLICK"] = "click";
    InteractionBrushSelectionStage["UN_DRAGGED"] = "unDragged";
    InteractionBrushSelectionStage["DRAGGED"] = "dragged";
})(InteractionBrushSelectionStage || (InteractionBrushSelectionStage = {}));
export var InteractionKeyboardKey;
(function (InteractionKeyboardKey) {
    InteractionKeyboardKey["SHIFT"] = "Shift";
    InteractionKeyboardKey["COPY"] = "c";
    InteractionKeyboardKey["ESC"] = "Escape";
    InteractionKeyboardKey["META"] = "Meta";
    InteractionKeyboardKey["CONTROL"] = "Control";
    InteractionKeyboardKey["ARROW_UP"] = "ArrowUp";
    InteractionKeyboardKey["ARROW_DOWN"] = "ArrowDown";
    InteractionKeyboardKey["ARROW_LEFT"] = "ArrowLeft";
    InteractionKeyboardKey["ARROW_RIGHT"] = "ArrowRight";
})(InteractionKeyboardKey || (InteractionKeyboardKey = {}));
export var SortMethodType;
(function (SortMethodType) {
    SortMethodType["ASC"] = "ASC";
    SortMethodType["DESC"] = "DESC";
})(SortMethodType || (SortMethodType = {}));
export var InterceptType;
(function (InterceptType) {
    InterceptType["HOVER"] = "hover";
    InterceptType["CLICK"] = "click";
    InterceptType["DATA_CELL_BRUSH_SELECTION"] = "dataCellBrushSelection";
    InterceptType["ROW_CELL_BRUSH_SELECTION"] = "rowCellBrushSelection";
    InterceptType["COL_CELL_BRUSH_SELECTION"] = "colCellBrushSelection";
    InterceptType["MULTI_SELECTION"] = "multiSelection";
    InterceptType["RESIZE"] = "resize";
})(InterceptType || (InterceptType = {}));
export const BRUSH_AUTO_SCROLL_INITIAL_CONFIG = {
    x: {
        value: 0,
        scroll: false,
    },
    y: {
        value: 0,
        scroll: false,
    },
};
export var ScrollbarPositionType;
(function (ScrollbarPositionType) {
    ScrollbarPositionType["CONTENT"] = "content";
    ScrollbarPositionType["CANVAS"] = "canvas";
})(ScrollbarPositionType || (ScrollbarPositionType = {}));
export var ScrollDirection;
(function (ScrollDirection) {
    ScrollDirection["SCROLL_UP"] = "scrollUp";
    ScrollDirection["SCROLL_DOWN"] = "scrollDown";
    ScrollDirection["SCROLL_LEFT"] = "scrollLeft";
    ScrollDirection["SCROLL_RIGHT"] = "scrollRight";
})(ScrollDirection || (ScrollDirection = {}));
export var ScrollDirectionRowIndexDiff;
(function (ScrollDirectionRowIndexDiff) {
    ScrollDirectionRowIndexDiff[ScrollDirectionRowIndexDiff["SCROLL_UP"] = -1] = "SCROLL_UP";
    ScrollDirectionRowIndexDiff[ScrollDirectionRowIndexDiff["SCROLL_DOWN"] = 1] = "SCROLL_DOWN";
})(ScrollDirectionRowIndexDiff || (ScrollDirectionRowIndexDiff = {}));
/**
 * 移动端滚动阈值（像素）
 * 用于区分用户的点击操作和滚动操作
 * 如果触摸移动距离小于此值，认为是点击；否则认为是滚动
 */
export const MOBILE_SCROLL_THRESHOLD = 10;
