// @antv/s2/esm/common/constant/resize.js
export const RESIZE_START_GUIDE_LINE_ID = 'RESIZE_START_GUIDE_LINE';
export const RESIZE_END_GUIDE_LINE_ID = 'RESIZE_END_GUIDE_LINE';
export const RESIZE_MASK_ID = 'RESIZE_MASK';
export const RESIZE_MIN_CELL_WIDTH = 40;
export const RESIZE_MIN_CELL_HEIGHT = 20;
// resize 时鼠标移动方向类型
export var ResizeDirectionType;
(function (ResizeDirectionType) {
    /**
     * 水平方向 resize
     */
    ResizeDirectionType["Horizontal"] = "col";
    /**
     * 垂直方向 resize
     */
    ResizeDirectionType["Vertical"] = "row";
})(ResizeDirectionType || (ResizeDirectionType = {}));
export var ResizeAreaEffect;
(function (ResizeAreaEffect) {
    ResizeAreaEffect["Field"] = "field";
    ResizeAreaEffect["Cell"] = "cell";
    ResizeAreaEffect["Tree"] = "tree";
    ResizeAreaEffect["Series"] = "series";
})(ResizeAreaEffect || (ResizeAreaEffect = {}));
export var ResizeType;
(function (ResizeType) {
    /**
     * 对所有单元格生效
     */
    ResizeType["ALL"] = "all";
    /**
     * 对当前单元格生效
     */
    ResizeType["CURRENT"] = "current";
    /**
     * 对当前单元格生效, 如果单元格是多选状态, 调整任意选中单元格, 对所有选中的生效
     */
    ResizeType["SELECTED"] = "selected";
})(ResizeType || (ResizeType = {}));
