// @antv/s2/esm/common/constant/events/basic.js
export var S2Event;
(function (S2Event) {
    /** ================ Row Cell ================  */
    S2Event["ROW_CELL_HOVER"] = "row-cell:hover";
    S2Event["ROW_CELL_CLICK"] = "row-cell:click";
    S2Event["ROW_CELL_DOUBLE_CLICK"] = "row-cell:double-click";
    S2Event["ROW_CELL_CONTEXT_MENU"] = "row-cell:context-menu";
    S2Event["ROW_CELL_MOUSE_DOWN"] = "row-cell:mouse-down";
    S2Event["ROW_CELL_MOUSE_UP"] = "row-cell:mouse-up";
    S2Event["ROW_CELL_MOUSE_MOVE"] = "row-cell:mouse-move";
    S2Event["ROW_CELL_SCROLL"] = "row-cell:scroll";
    S2Event["ROW_CELL_BRUSH_SELECTION"] = "row-cell:brush-selection";
    S2Event["ROW_CELL_COLLAPSED"] = "row-cell:collapsed";
    S2Event["ROW_CELL_ALL_COLLAPSED"] = "row-cell:all-collapsed";
    S2Event["ROW_CELL_RENDER"] = "row-cell:render";
    S2Event["ROW_CELL_SELECTED"] = "row-cell:selected";
    // 内部用来通信的 event
    S2Event["ROW_CELL_COLLAPSED__PRIVATE"] = "row-cell:collapsed__private";
    S2Event["ROW_CELL_ALL_COLLAPSED__PRIVATE"] = "row-cell:all-collapsed__private";
    S2Event["DATA_CELL_HOVER_TRIGGERED_PRIVATE"] = "data-cell:hover-trigger__private";
    S2Event["DATA_CELL_CLICK_TRIGGERED_PRIVATE"] = "data-cell:click-trigger__private";
    /** ================ Col Cell ================  */
    S2Event["COL_CELL_HOVER"] = "col-cell:hover";
    S2Event["COL_CELL_CLICK"] = "col-cell:click";
    S2Event["COL_CELL_DOUBLE_CLICK"] = "col-cell:double-click";
    S2Event["COL_CELL_CONTEXT_MENU"] = "col-cell:context-menu";
    S2Event["COL_CELL_MOUSE_DOWN"] = "col-cell:mouse-down";
    S2Event["COL_CELL_MOUSE_UP"] = "col-cell:mouse-up";
    S2Event["COL_CELL_MOUSE_MOVE"] = "col-cell:mouse-move";
    S2Event["COL_CELL_BRUSH_SELECTION"] = "col-cell:brush-selection";
    S2Event["COL_CELL_EXPANDED"] = "col-cell:expanded";
    S2Event["COL_CELL_EXPAND_ICON_HOVER"] = "col-cell:expand-icon-hover";
    S2Event["COL_CELL_HIDDEN"] = "col-cell:hidden";
    S2Event["COL_CELL_RENDER"] = "col-cell:render";
    S2Event["COL_CELL_SELECTED"] = "col-cell:selected";
    /** ================ Data Cell ================  */
    S2Event["DATA_CELL_HOVER"] = "data-cell:hover";
    S2Event["DATA_CELL_CLICK"] = "data-cell:click";
    S2Event["DATA_CELL_DOUBLE_CLICK"] = "data-cell:double-click";
    S2Event["DATA_CELL_CONTEXT_MENU"] = "data-cell:context-menu";
    S2Event["DATA_CELL_MOUSE_UP"] = "data-cell:mouse-up";
    S2Event["DATA_CELL_MOUSE_DOWN"] = "data-cell:mouse-down";
    S2Event["DATA_CELL_MOUSE_MOVE"] = "data-cell:mouse-move";
    S2Event["DATA_CELL_BRUSH_SELECTION"] = "data-cell:brush-selection";
    S2Event["DATA_CELL_SELECT_MOVE"] = "data-cell:select-move";
    S2Event["DATA_CELL_RENDER"] = "data-cell:render";
    S2Event["DATA_CELL_SELECTED"] = "data-cell:selected";
    /** ================ Corner Cell ================  */
    S2Event["CORNER_CELL_HOVER"] = "corner-cell:hover";
    S2Event["CORNER_CELL_CLICK"] = "corner-cell:click";
    S2Event["CORNER_CELL_DOUBLE_CLICK"] = "corner-cell:double-click";
    S2Event["CORNER_CELL_CONTEXT_MENU"] = "corner-cell:context-menu";
    S2Event["CORNER_CELL_MOUSE_DOWN"] = "corner-cell:mouse-down";
    S2Event["CORNER_CELL_MOUSE_UP"] = "corner-cell:mouse-up";
    S2Event["CORNER_CELL_MOUSE_MOVE"] = "corner-cell:mouse-move";
    S2Event["CORNER_CELL_RENDER"] = "corner-cell:render";
    S2Event["CORNER_CELL_SELECTED"] = "corner-cell:selected";
    /** ================ Merged Cells ================  */
    S2Event["MERGED_CELLS_HOVER"] = "merged-cells:hover";
    S2Event["MERGED_CELLS_CLICK"] = "merged-cells:click";
    S2Event["MERGED_CELLS_DOUBLE_CLICK"] = "merged-cells:double-click";
    S2Event["MERGED_CELLS_CONTEXT_MENU"] = "merged-cells:context-menu";
    S2Event["MERGED_CELLS_MOUSE_DOWN"] = "merged-cells:mouse-down";
    S2Event["MERGED_CELLS_MOUSE_UP"] = "merged-cells:mouse-up";
    S2Event["MERGED_CELLS_MOUSE_MOVE"] = "merged-cells:mouse-move";
    S2Event["MERGED_CELLS_RENDER"] = "merged-cells:render";
    /** ================ SeriesNumber Cell ================  */
    S2Event["SERIES_NUMBER_CELL_RENDER"] = "series-number-cell:render";
    /** ================ Sort ================  */
    S2Event["RANGE_SORT"] = "sort:range-sort";
    S2Event["RANGE_SORTED"] = "sort:range-sorted";
    /** ================ Table Filter ================  */
    S2Event["RANGE_FILTER"] = "filter:range-filter";
    S2Event["RANGE_FILTERED"] = "filter:range-filtered";
    /** ================ Table Layout ================  */
    S2Event["LAYOUT_AFTER_HEADER_LAYOUT"] = "layout:after-header-layout";
    S2Event["LAYOUT_CELL_RENDER"] = "layout:cell-render";
    S2Event["LAYOUT_PAGINATION"] = "layout:pagination";
    S2Event["LAYOUT_AFTER_REAL_DATA_CELL_RENDER"] = "layout:after-real-data-cell-render";
    S2Event["LAYOUT_AFTER_RENDER"] = "layout:after-render";
    S2Event["LAYOUT_BEFORE_RENDER"] = "layout:before-render";
    S2Event["LAYOUT_DESTROY"] = "layout:destroy";
    /** ================ Table Layout Resize ================  */
    S2Event["LAYOUT_RESIZE"] = "layout:resize";
    S2Event["LAYOUT_RESIZE_SERIES_WIDTH"] = "layout:resize-series-width";
    S2Event["LAYOUT_RESIZE_ROW_WIDTH"] = "layout:resize-row-width";
    S2Event["LAYOUT_RESIZE_ROW_HEIGHT"] = "layout:resize-row-height";
    S2Event["LAYOUT_RESIZE_COL_WIDTH"] = "layout:resize-column-width";
    S2Event["LAYOUT_RESIZE_COL_HEIGHT"] = "layout:resize-column-height";
    S2Event["LAYOUT_RESIZE_TREE_WIDTH"] = "layout:resize-tree-width";
    S2Event["LAYOUT_RESIZE_MOUSE_DOWN"] = "layout:resize:mouse-down";
    S2Event["LAYOUT_RESIZE_MOUSE_MOVE"] = "layout:resize:mouse-move";
    S2Event["LAYOUT_RESIZE_MOUSE_UP"] = "layout:resize-mouse-up";
    /** ================ Global ================  */
    S2Event["GLOBAL_KEYBOARD_DOWN"] = "global:keyboard-down";
    S2Event["GLOBAL_KEYBOARD_UP"] = "global:keyboard-up";
    S2Event["GLOBAL_COPIED"] = "global:copied";
    S2Event["GLOBAL_MOUSE_UP"] = "global:mouse-up";
    S2Event["GLOBAL_MOUSE_MOVE"] = "global:mouse-move";
    S2Event["GLOBAL_ACTION_ICON_CLICK"] = "global:action-icon-click";
    S2Event["GLOBAL_ACTION_ICON_HOVER"] = "global:action-icon-hover";
    S2Event["GLOBAL_ACTION_ICON_HOVER_OFF"] = "global:action-icon-hover-off";
    S2Event["GLOBAL_CONTEXT_MENU"] = "global:context-menu";
    S2Event["GLOBAL_CLICK"] = "global:click";
    S2Event["GLOBAL_PREVIEW_CLICK"] = "global:preview-click";
    S2Event["GLOBAL_DOUBLE_CLICK"] = "global:double-click";
    S2Event["GLOBAL_SELECTED"] = "global:selected";
    S2Event["GLOBAL_HOVER"] = "global:hover";
    S2Event["GLOBAL_RESET"] = "global:reset";
    S2Event["GLOBAL_LINK_FIELD_JUMP"] = "global:link-field-jump";
    S2Event["GLOBAL_SCROLL"] = "global:scroll";
})(S2Event || (S2Event = {}));
