// @antv/s2/lib/common/constant/basic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_TEXT_LINE_HEIGHT = exports.ELLIPSIS_SYMBOL = exports.getDefaultCornerText = exports.getDefaultSeriesNumberText = exports.MiniChartType = exports.LAYOUT_SAMPLE_COUNT = exports.DEFAULT_CORNER_MAX_WIDTH_RATIO = exports.PRECISION = exports.EMPTY_PLACEHOLDER = exports.HORIZONTAL_RESIZE_AREA_KEY_PRE = exports.KEY_GROUP_EMPTY_PLACEHOLDER = exports.KEY_GROUP_FROZEN_SPLIT_LINE = exports.KEY_GROUP_CORNER_RESIZE_AREA = exports.KEY_GROUP_CORNER_SCROLL = exports.KEY_GROUP_COL_RESIZE_AREA = exports.KEY_GROUP_COL_FROZEN_TRAILING = exports.KEY_GROUP_COL_FROZEN = exports.KEY_GROUP_COL_SCROLL = exports.KEY_GROUP_ROW_RESIZE_AREA = exports.KEY_GROUP_ROW_FROZEN_TRAILING = exports.KEY_GROUP_ROW_FROZEN = exports.KEY_GROUP_ROW_SCROLL = exports.KEY_GROUP_ROW_INDEX_RESIZE_AREA = exports.KEY_GROUP_ROW_INDEX_FROZEN_TRAILING = exports.KEY_GROUP_ROW_INDEX_FROZEN = exports.KEY_GROUP_ROW_INDEX_SCROLL = exports.KEY_GROUP_MERGED_CELLS = exports.KEY_GROUP_PANEL_SCROLL = exports.KEY_GROUP_GRID_GROUP = exports.KEY_GROUP_PANEL_GROUND = exports.KEY_GROUP_FORE_GROUND = exports.KEY_GROUP_BACK_GROUND = exports.EMPTY_PLACEHOLDER_GROUP_CONTAINER_Z_INDEX = exports.PANEL_GROUP_FROZEN_GROUP_Z_INDEX = exports.PANEL_MERGE_GROUP_Z_INDEX = exports.PANEL_GRID_GROUP_Z_INDEX = exports.PANEL_GROUP_SCROLL_GROUP_Z_INDEX = exports.PANEL_GROUP_GROUP_CONTAINER_Z_INDEX = exports.FRONT_GROUND_GROUP_BRUSH_SELECTION_Z_INDEX = exports.FRONT_GROUND_GROUP_RESIZE_AREA_Z_INDEX = exports.FRONT_GROUND_GROUP_FROZEN_Z_INDEX = exports.FRONT_GROUND_GROUP_SCROLL_Z_INDEX = exports.FRONT_GROUND_GROUP_CONTAINER_Z_INDEX = exports.BACK_GROUND_GROUP_CONTAINER_Z_INDEX = void 0;
const i18n_1 = require("../i18n");
// 约定这个 z-index 为 0 的 container 作为基准
exports.BACK_GROUND_GROUP_CONTAINER_Z_INDEX = 0;
// foregroundGroup 上的 children 层叠顺序
exports.FRONT_GROUND_GROUP_CONTAINER_Z_INDEX = 3;
exports.FRONT_GROUND_GROUP_SCROLL_Z_INDEX = 3;
exports.FRONT_GROUND_GROUP_FROZEN_Z_INDEX = 4;
exports.FRONT_GROUND_GROUP_RESIZE_AREA_Z_INDEX = 5;
exports.FRONT_GROUND_GROUP_BRUSH_SELECTION_Z_INDEX = 5;
// panelGroup 上的 children 层叠顺序
exports.PANEL_GROUP_GROUP_CONTAINER_Z_INDEX = 1;
exports.PANEL_GROUP_SCROLL_GROUP_Z_INDEX = 1;
exports.PANEL_GRID_GROUP_Z_INDEX = 2;
exports.PANEL_MERGE_GROUP_Z_INDEX = 3;
exports.PANEL_GROUP_FROZEN_GROUP_Z_INDEX = 4;
// emptyPlaceholderGroup 上的 children 层叠顺序
exports.EMPTY_PLACEHOLDER_GROUP_CONTAINER_Z_INDEX = 4;
// group's key
exports.KEY_GROUP_BACK_GROUND = 'backGroundGroup';
exports.KEY_GROUP_FORE_GROUND = 'foreGroundGroup';
exports.KEY_GROUP_PANEL_GROUND = 'panelGroup';
exports.KEY_GROUP_GRID_GROUP = 'gridGroup';
exports.KEY_GROUP_PANEL_SCROLL = 'panelScrollGroup';
exports.KEY_GROUP_MERGED_CELLS = 'mergedCellsGroup';
/**
 * series number
 */
exports.KEY_GROUP_ROW_INDEX_SCROLL = 'rowIndexScrollGroup';
exports.KEY_GROUP_ROW_INDEX_FROZEN = 'rowIndexFrozenGroup';
exports.KEY_GROUP_ROW_INDEX_FROZEN_TRAILING = 'rowIndexFrozenTrailingGroup';
exports.KEY_GROUP_ROW_INDEX_RESIZE_AREA = 'rowIndexResizeAreaGroup';
/**
 * row
 */
exports.KEY_GROUP_ROW_SCROLL = 'rowScrollGroup';
exports.KEY_GROUP_ROW_FROZEN = 'rowHeaderFrozenGroup';
exports.KEY_GROUP_ROW_FROZEN_TRAILING = 'rowHeaderFrozenTrailingGroup';
exports.KEY_GROUP_ROW_RESIZE_AREA = 'rowResizeAreaGroup';
/**
 * column
 */
exports.KEY_GROUP_COL_SCROLL = 'colScrollGroup';
exports.KEY_GROUP_COL_FROZEN = 'colFrozenGroup';
exports.KEY_GROUP_COL_FROZEN_TRAILING = 'colFrozenTrailingGroup';
exports.KEY_GROUP_COL_RESIZE_AREA = 'colResizeAreaGroup';
/**
 * corner
 */
exports.KEY_GROUP_CORNER_SCROLL = 'cornerScrollGroup';
exports.KEY_GROUP_CORNER_RESIZE_AREA = 'cornerResizeAreaGroup';
exports.KEY_GROUP_FROZEN_SPLIT_LINE = 'frozenSplitLine';
exports.KEY_GROUP_EMPTY_PLACEHOLDER = 'emptyPlaceholderGroup';
exports.HORIZONTAL_RESIZE_AREA_KEY_PRE = 'horizontal-resize-area-';
exports.EMPTY_PLACEHOLDER = '-';
// data precision
exports.PRECISION = 16;
// 角头最大占整个容器的比例 (0-1)
exports.DEFAULT_CORNER_MAX_WIDTH_RATIO = 0.5;
/** 布局采样数 */
exports.LAYOUT_SAMPLE_COUNT = 50;
/** mini 图类型 */
var MiniChartType;
(function (MiniChartType) {
    MiniChartType["Line"] = "line";
    MiniChartType["Bar"] = "bar";
    MiniChartType["Bullet"] = "bullet";
})(MiniChartType || (exports.MiniChartType = MiniChartType = {}));
const getDefaultSeriesNumberText = () => (0, i18n_1.i18n)('序号');
exports.getDefaultSeriesNumberText = getDefaultSeriesNumberText;
const getDefaultCornerText = () => (0, i18n_1.i18n)('指标');
exports.getDefaultCornerText = getDefaultCornerText;
// 省略号
exports.ELLIPSIS_SYMBOL = '...';
exports.DEFAULT_TEXT_LINE_HEIGHT = 16;
