// @antv/s2/lib/common/constant/frozen.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FrozenGroupAreaTypeMap = exports.FrozenGroupType = exports.FrozenGroupArea = void 0;
var FrozenGroupArea;
(function (FrozenGroupArea) {
    FrozenGroupArea["Row"] = "frozenRow";
    FrozenGroupArea["Col"] = "frozenCol";
    FrozenGroupArea["TrailingRow"] = "frozenTrailingRow";
    FrozenGroupArea["TrailingCol"] = "frozenTrailingCol";
})(FrozenGroupArea || (exports.FrozenGroupArea = FrozenGroupArea = {}));
var FrozenGroupType;
(function (FrozenGroupType) {
    FrozenGroupType["Row"] = "frozenRowGroup";
    FrozenGroupType["Col"] = "frozenColGroup";
    FrozenGroupType["TrailingCol"] = "frozenTrailingColGroup";
    FrozenGroupType["TrailingRow"] = "frozenTrailingRowGroup";
    FrozenGroupType["TopLeft"] = "frozenTopLeftGroup";
    FrozenGroupType["TopRight"] = "frozenTopRightGroup";
    FrozenGroupType["BottomLeft"] = "frozenBottomLeftGroup";
    FrozenGroupType["BottomRight"] = "frozenBottomRightGroup";
    FrozenGroupType["Scroll"] = "scrollGroup";
})(FrozenGroupType || (exports.FrozenGroupType = FrozenGroupType = {}));
exports.FrozenGroupAreaTypeMap = {
    [FrozenGroupArea.Row]: FrozenGroupType.Row,
    [FrozenGroupArea.Col]: FrozenGroupType.Col,
    [FrozenGroupArea.TrailingRow]: FrozenGroupType.TrailingRow,
    [FrozenGroupArea.TrailingCol]: FrozenGroupType.TrailingCol,
};
