// @antv/s2/lib/common/constant/theme.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CELL_PADDING = exports.G2_THEME_TYPE = exports.DARK_THEME_CLS = exports.INTERVAL_BAR_HEIGHT = exports.FONT_FAMILY = exports.PALETTE_MAP = void 0;
const colorful_1 = require("../../theme/palette/colorful");
const dark_1 = require("../../theme/palette/dark");
const default_1 = require("../../theme/palette/default");
const gray_1 = require("../../theme/palette/gray");
const classnames_1 = require("./classnames");
exports.PALETTE_MAP = {
    default: default_1.paletteDefault,
    colorful: colorful_1.paletteColorful,
    gray: gray_1.paletteGray,
    dark: dark_1.paletteDark,
};
/**
 * 默认字体
 *
 * -apple-system 会导致 iOS15 崩溃
 * BlinkMacSystemFont 会导致 g5.0 下 jest-electron 测试环境崩溃
 */
exports.FONT_FAMILY = 'Roboto, PingFangSC, Microsoft YaHei, Arial, sans-serif';
exports.INTERVAL_BAR_HEIGHT = 12;
exports.DARK_THEME_CLS = `${classnames_1.S2_PREFIX_CLS}-dark-theme`;
/**
 * 兼容 G2 主题: S2 和 G2 的主题名转换
 * https://g2.antv.antgroup.com/manual/core/theme
 */
exports.G2_THEME_TYPE = {
    default: 'light',
    colorful: 'light',
    gray: 'light',
    dark: 'dark',
};
exports.CELL_PADDING = 8;
