// @antv/s2/lib/common/constant/tooltip.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTooltipOperatorTableSortMenus = exports.getTooltipOperatorSortMenus = exports.getTooltipOperatorHiddenColumnsMenu = exports.TOOLTIP_POSITION_OFFSET = exports.TOOLTIP_OPERATION_PREFIX_CLS = exports.TOOLTIP_CONTAINER_HIDE_CLS = exports.TOOLTIP_CONTAINER_SHOW_CLS = exports.TOOLTIP_CONTAINER_CLS = exports.MOBILE_TOOLTIP_PREFIX_CLS = exports.TOOLTIP_PREFIX_CLS = void 0;
const classnames_1 = require("../../common/constant/classnames");
const i18n_1 = require("../../common/i18n");
exports.TOOLTIP_PREFIX_CLS = `${classnames_1.S2_PREFIX_CLS}-tooltip`;
exports.MOBILE_TOOLTIP_PREFIX_CLS = `${exports.TOOLTIP_PREFIX_CLS}-mobile`;
exports.TOOLTIP_CONTAINER_CLS = `${exports.TOOLTIP_PREFIX_CLS}-container`;
exports.TOOLTIP_CONTAINER_SHOW_CLS = `${exports.TOOLTIP_CONTAINER_CLS}-show`;
exports.TOOLTIP_CONTAINER_HIDE_CLS = `${exports.TOOLTIP_CONTAINER_CLS}-hide`;
exports.TOOLTIP_OPERATION_PREFIX_CLS = `${exports.TOOLTIP_PREFIX_CLS}-operation`;
exports.TOOLTIP_POSITION_OFFSET = {
    x: 15,
    y: 10,
};
const getTooltipOperatorHiddenColumnsMenu = () => {
    return {
        key: 'hiddenColumns',
        label: (0, i18n_1.i18n)('隐藏'),
        icon: 'EyeOutlined',
    };
};
exports.getTooltipOperatorHiddenColumnsMenu = getTooltipOperatorHiddenColumnsMenu;
const getTooltipOperatorSortMenus = () => [
    {
        key: 'asc',
        icon: 'groupAsc',
        label: (0, i18n_1.i18n)('组内升序'),
    },
    {
        key: 'desc',
        icon: 'groupDesc',
        label: (0, i18n_1.i18n)('组内降序'),
    },
    {
        key: 'none',
        label: (0, i18n_1.i18n)('不排序'),
    },
];
exports.getTooltipOperatorSortMenus = getTooltipOperatorSortMenus;
const getTooltipOperatorTableSortMenus = () => [
    {
        key: 'asc',
        icon: 'groupAsc',
        label: (0, i18n_1.i18n)('升序'),
    },
    {
        key: 'desc',
        icon: 'groupDesc',
        label: (0, i18n_1.i18n)('降序'),
    },
    {
        key: 'none',
        label: (0, i18n_1.i18n)('不排序'),
    },
];
exports.getTooltipOperatorTableSortMenus = getTooltipOperatorTableSortMenus;
